### PR TITLE
[Ming-Omni] Support Video Understanding for V1 Pipeline

### PR DESCRIPTION
**Pull request by yuan-luo**
## Summary

Adds video understanding support to the Ming-Omni V1 pipeline. Videos flow through the existing `image_encoder` stage alongside images producing `video_embeds` that the thinker model runner injects at `video_patch_token` positions.

No new stage or process is added. `MingOmniVisionEncoder` already handles temporal patches via Conv3D, and `ming_thinker_model_runner` already routes `video_embeds` by token id; this PR fills in the missing preprocessor / encoder forward / merge plumbing.

## Modifications

**`sglang_omni_v1/models/ming_omni/components/preprocessor.py`**
- Add `VIDEO_START` / `VIDEO_END` / `VIDEO_PATCH` modality tokens.
- Add `_inject_top_level_videos`, mirroring `_inject_top_level_images`, so requests can pass `videos=[...]` at the top level or as inline `video_url` items.
- Resolve `_video_patch_id` from `llm_config.video_patch_token` with a tokenizer fallback.
- Extract `videos` from raw inputs and async-load via `ensure_video_list_async` together with images and audio.
- Add `_process_videos`, which calls `Qwen2VLImageProcessor.preprocess(images=None, videos=...)` and returns `pixel_values_videos` + `video_grid_thw` + token counts. The `images` positional has to be passed explicitly because `BaseImageProcessor.__call__` requires it.
- Populate `encoder_inputs[IMAGE_STAGE]` with `pixel_values_videos` / `video_grid_thw` alongside the image tensors. Combined cache key uses `img:.../vid:...` namespacing so prefix-cache reuse is correct for either modality alone or both together.
- Extend `_build_prompt` to handle `video_url` / `video` content items and emit `<videoPatch>` placeholders at the right positions.

**`sglang_omni_v1/models/ming_omni/components/image_encoder.py`**
- Factor the encoder + projector + L2-normalize path into a private `_encode(pixel_values, grid_thw)` helper so images and videos share the exact same compute.
- `forward()` now accepts optional `pixel_values_videos` / `video_grid_thw` kwargs (all four image/video inputs default to `None`). It returns `image_embeds` / `video_embeds` / token counts only for the modalities that were provided, so the same factory works for images-only, videos-only, or both.

**`sglang_omni_v1/models/ming_omni/pipeline/merge.py`**
- `build_thinker_inputs` extracts `video_embeds` from the image-stage output and injects it into `thinker_model_inputs` next to `image_embeds` / `audio_embeds`. The thinker runner already knows how to dispatch by token id.
- `media_cache_keys["image"]` carries the combined image/video namespace produced by the preprocessor.

## Tests

Three new unit tests in `tests/unit_test/ming_omni/test_pipeline.py`:

- `test_ming_preprocessor_injects_top_level_videos_as_inline_content` — `_inject_top_level_videos` produces the right inline `video_url` item, leaves the system message alone, and does not mutate the caller's list.
- `test_ming_image_encoder_forward_accepts_video_inputs` — `MingImageEncoder.forward` signature exposes `pixel_values_videos` / `video_grid_thw`; all four image/video kwargs default to `None`.
- `test_ming_merge_extracts_video_embeds_into_thinker_inputs` — `build_thinker_inputs` routes `video_embeds` from the image stage into `thinker_model_inputs` with shape preserved and `media_cache_keys` namespaced correctly.

All three pass on the H200 dev box.

## End-to-end validation

Launched the speech server with `--tp-size 2 --gpu-thinker 2 --gpu-talker 0 --mem-fraction-static 0.80` against `inclusionAI/Ming-flash-omni-2.0` on H200 and sent:

```json
{
  "model": "ming-omni",
  "messages": [
    {"role": "user", "content": "请简短描述这段视频里发生了什么。"}
  ],
  "videos": ["/sgl-workspace/sglang-omni-dev/tests/data/draw.mp4"],
  "video_max_frames": 16,
  "max_tokens": 200
}
```

Model output (excerpt):

> 这段视频呈现了**以俯视视角，拍摄一位身穿白色长袖衣物的人，手持一支白色触控笔，在平板电脑（如iPad）的屏幕上进行数字绘画**的场景... 屏幕上正在绘制的是一个**乐器（吉他）的轮廓图**...

Hits every keyword that the existing Qwen3-Omni docs test (`tests/docs/qwen3_omni/test_docs_qwen3_omni.py::EXPECTED_VIDEO_KEYWORDS = ["draw", "stylus", "pen", "tablet", "girl"]`) expects.

## Request formats supported

Top-level (Ming/Qwen3-Omni convention):

```json
{
  "model": "ming-omni",
  "messages": [{"role": "user", "content": "Describe the video"}],
  "videos": ["/path/to/clip.mp4"],
  "video_max_frames": 16
}
```

Inline content (OpenAI-style multimodal content array):

```json
{
  "messages": [{
    "role": "user",
    "content": [
      {"type": "video_url", "video_url": {"url": "/path/to/clip.mp4"}},
      {"type": "text", "text": "Describe the video"}
    ]
  }]
}
```

## Test plan

- [ ] `pytest tests/unit_test/ming_omni/test_pipeline.py -k video -v`
- [ ] Launch `examples/run_ming_omni_speech_server.py --version v1 ...` and send a request with `videos: ["..."]`
- [ ] Verify response references the video content (drawing / stylus / tablet for `draw.mp4`)
- [ ] Verify image-only and audio-only requests still work (no regression on the existing modalities)

## Note

This PR is based on V1 migration of Ming https://github.com/sgl-project/sglang-omni/pull/437. Need to be merged after #437.

### CODE REVIEW

File: sglang_omni/models/ming_omni/pipeline/merge.py [hunk 1]
```diff
@@ -102,11 +107,18 @@ def build_thinker_inputs(
...
+        thinker_model_inputs["video_embeds"] = video_embeds
+
     media_cache_keys: dict[str, str] = {}
     encoder_inputs = state.encoder_inputs or {}
     image_ck = (encoder_inputs.get(IMAGE_STAGE) or {}).get("cache_key")
     audio_ck = (encoder_inputs.get(AUDIO_STAGE) or {}).get("cache_key")
     if image_ck:
+        # image_ck already namespaces image vs video (img:.../vid:...) when both
+        # are present so we keep one cache key for the combined visual stage.
         media_cache_keys["image"] = f"image:{image_ck}"
```
Comment (edwingao28) [inline sglang_omni/models/ming_omni/pipeline/merge.py:126]: Blocker: video patch tokens are never de-aliased for prefix caching. This only writes `media_cache_keys["image"]`, but the SGLang adapter looks up keys by modality (`image`, `audio`, `video`). As a result `media_cache_keys.get("video")` is `None`, so `video_patch_token` remains the raw token id.

Two different videos with the same number of placeholders can therefore alias in the radix prefix cache. Please carry image/video cache keys separately into `media_cache_keys`, even if the encoder stage itself still uses one combined visual-stage cache key.

File: sglang_omni/models/ming_omni/components/preprocessor.py [hunk 2]
```diff
@@ -280,16 +373,50 @@ async def __call__(self, payload: StagePayload) -> StagePayload:
...
+                            if vid:
+                                raw_videos.append(vid)
 
         # Compute cache keys BEFORE async loading; same content -> same key so
         # SGLang's radix prefix cache can correctly reuse KVs across requests, and
         # different content -> different key so it never falsely aliases image
         # placeholder positions (which share the same generic image_patch_token).
         image_cache_key = compute_image_cache_key(raw_images) if raw_images else None
         audio_cache_key = compute_audio_cache_key(audio_urls) if audio_urls else None
+        video_cache_key = compute_video_cache_key(raw_videos) if raw_videos else None
```
Comment (edwingao28) [inline sglang_omni/models/ming_omni/components/preprocessor.py]: Blocker: the video cache key hashes the raw media identity, but not the decode parameters. The same video can produce different frame counts under different `video_fps`, `video_max_frames`, or pixel limits, while this key stays identical.

That can make the encoder output cache return `video_embeds` with a length that no longer matches the prompt placeholders. Please include the effective video decode params in the video cache key.

File: sglang_omni/models/ming_omni/components/preprocessor.py [hunk 3]
```diff
@@ -540,6 +705,18 @@ def append_placeholder(
                                 n_tokens,
                             )
                             image_idx += 1
+                        elif item_type in ("video_url", "video"):
+                            n_tokens = (
+                                v_counts[video_idx] if video_idx < len(v_counts) else 1
```
Comment (edwingao28) [inline sglang_omni/models/ming_omni/components/preprocessor.py:732]: Note: If video loading fails, `video_token_counts` is empty but the original `video_url` item is still in `messages`, so this fallback emits one orphan `video_patch_token`. The request then has a video placeholder with no corresponding `video_embeds`.

File: sglang_omni/models/ming_omni/components/preprocessor.py [hunk 4]
```diff
@@ -237,6 +276,45 @@ def _process_images(
...
+        """
+        processor = self._get_image_processor()
+        # Convert per-video tensors to numpy arrays in (T, H, W, C) uint8 — the
+        # format Qwen2VLImageProcessor expects when ``videos`` is a list of
+        # per-video frame stacks.
+        np_videos: list[np.ndarray] = []
+        for v in videos:
+            t = v
+            if isinstance(t, torch.Tensor):
+                arr = t.detach().cpu().to(torch.uint8).numpy()
```
Comment (edwingao28) [inline sglang_omni/models/ming_omni/components/preprocessor.py]: Small correctness fix: `load_video_path` resizes with BICUBIC + antialias and returns a float tensor. BICUBIC can overshoot outside `[0, 255]`, and `to(torch.uint8)` wraps instead of saturating.

Please clamp before casting, e.g. `t.detach().cpu().clamp_(0, 255).to(torch.uint8).numpy()`.

File: tests/unit_test/ming_omni/test_pipeline.py [hunk 5]
```diff
@@ -460,3 +460,100 @@ def from_server_args(cls, **kwargs):
...
+    from sglang_omni.models.ming_omni.components.image_encoder import MingImageEncoder
+
+    params = inspect.signature(MingImageEncoder.forward).parameters
+    for name in (
+        "pixel_values",
+        "image_grid_thw",
+        "pixel_values_videos",
+        "video_grid_thw",
+    ):
+        assert name in params, f"missing forward kwarg: {name}"
```
Comment (edwingao28) [inline tests/unit_test/ming_omni/test_pipeline.py:535]: This test only checks the `forward` signature, so it would still pass if video encoding returned the wrong shape or ignored videos entirely.

Please add at least one real video forward-path test that asserts `video_embeds.shape[0] == sum(video_token_counts)`, plus regression coverage for image+video together and the decode-failure path.

**Review by edwingao28 — CHANGES_REQUESTED** [review]
Requesting changes for two correctness blockers in video caching.

1. Video patch tokens are not de-aliased because `media_cache_keys["video"]` is never populated, so different videos with the same placeholder length can alias in the radix prefix cache.
2. The video encoder cache key does not include decode params such as `video_fps`, `video_max_frames`, and pixel limits, so the same video can hit stale embeds with the wrong token count.

I also left smaller comments for decode-failure placeholder handling, uint8 clamping, and missing regression coverage.

File: sglang_omni/models/ming_omni/pipeline/merge.py (hunk 1 above)
Comment (yuan-luo) [reply sglang_omni/models/ming_omni/pipeline/merge.py:126]: Resolved.

File: sglang_omni/models/ming_omni/components/preprocessor.py (hunk 2 above)
Comment (yuan-luo) [reply sglang_omni/models/ming_omni/components/preprocessor.py]: Addressed.

File: sglang_omni/models/ming_omni/components/preprocessor.py (hunk 3 above)
Comment (yuan-luo) [reply sglang_omni/models/ming_omni/components/preprocessor.py:732]: Addressed.

File: sglang_omni/models/ming_omni/components/preprocessor.py (hunk 4 above)
Comment (yuan-luo) [reply sglang_omni/models/ming_omni/components/preprocessor.py]: Fixed.

File: tests/unit_test/ming_omni/test_pipeline.py (hunk 5 above)
Comment (yuan-luo) [reply tests/unit_test/ming_omni/test_pipeline.py:535]: Done. Added three behavioral tests covering all three requested cases.

**Review by edwingao28 — APPROVED** [review]
